### ISSUE TITLE
Expand `derive` documentation.

### ISSUE THREAD

**Issue by ehuss** [rust-lang-nursery/reference#540]
The `derive` attribute documentation should probably list the built-in traits that are derivable. To my knowledge, the list is:

- `Clone`
- `Hash`
- `PartialEq`
- `Eq`
- `PartialOrd`
- `Ord`
- `Debug`
- `Default`
- `Send`
- `Sync`
- `Copy`

Should the documentation also specify the semantics of how the default implementations behave? It seems like important information, but I think it is mostly covered in the core/std docs. Another option is to link to https://doc.rust-lang.org/book/appendix-03-derivable-traits.html.

**Comment by Mark-Simulacrum** [comment]
Send and Sync to my knowledge can't be derived -- they're auto-implemented by the compiler where applicable and are otherwise `unsafe` to implement.

**Comment by ehuss** [comment]
Oops, I had just copy/pasted from [the code](https://github.com/rust-lang/rust/blob/2c8bbf50db0ef90a33f986ba8fc2e1fe129197ff/src/libsyntax_ext/deriving/mod.rs#L65-L90) and didn't pay attention.

**Comment by Centril** [comment]
> Should the documentation also specify the semantics of how the default implementations behave? It seems like important information, but I think it is mostly covered in the core/std docs.

Yes, the reference should strive to be the source of truth to which everything links, not the other way around. (caveat for things about rustc and notes...) The info should be here.

**Comment by lcnr** [comment]
We should also document:
- the standard library `derive` also looks at fields and adds bounds for `T::Assoc`, but not `<T as Trait>::Assoc` https://github.com/lcnr/random-rust-snippets/issues/8
- to do this, it does not allow macro's in type position https://github.com/lcnr/random-rust-snippets/issues/9
